File: c/daa/quicksort.c
#include<stdio.h>
int partition(int* a,int p,int r){
	int x=a[r];
	int i=p-1;
	int j,temp;
	for(j=p;j<r;j++){
		if(a[j]<=x){
			i +=1;
			temp=a[i];
			a[i]=a[j];
			a[j]=temp;
		}
	}
	temp=a[i+1];
	a[i+1]=a[r];
	a[r]=temp;
	return i+1;
}
void quicksort(int* a,int p,int r){
	if (p<r){
		int q=partition(a,p,r);
		quicksort(a,p,q-1);
		quicksort(a,q+1,r);
	}
}
void printarray(int* a,int n){
	int i;
	for(i=0;i<n;i++){
		printf("%d ",a[i]);
	}
	printf("\n");
}
int main(int args,char** argv){
	int arr[]={-5,-3,-4,7,9,2,1,3,4};
	int length= sizeof(arr)/sizeof(arr[0]);
	printf("given array: ");
	printarray(arr,length);
	quicksort(arr,0,length);
	printf("sorted array: ");
	printarray(arr,length);
	return 0;
}

File: c/daa/daa_in_python/quicksort.py
def partition(A,p,r):
	x=A[p]
	i=p
	j=r
	while i<j:
		while i<j and A[j]>x:
			j=j-1
		while i<j and A[i]<=x:
			i=i+1
		if i<j:
			A[i],A[j]=A[j],A[i]
	A[p],A[j]=A[j],A[p]
	return j
def quicksort(A,p,r):
	if p<r:
		q=partition(A,p,r)
		quicksort(A,p,q-1)
		quicksort(A,q+1,r)
arr = [5, 2, 9, 1, 5, 6]
print("Original array:", arr)
quicksort(arr,0,len(arr)-1)
print("Sorted array:", arr)
